### DISCUSSION TITLE
Few questions about Avalonia MacOS

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#11925] [Q&A] [answered]
Hi all,
I just tried the Avalonia app I am working on for a few months on MacOS system.
For the context, I am developing on Windows and tried the app periodically on WASM, and I have no experience at all with MacOS... 

First of all, well done guys for Avalonia! I installed VS for Mac, unziped my project, loaded the solution, clicked on the Play arrow and voilà, it worked the first time!!! :)) 

Here are the few remarks and questions I have about this first experience:

1) Compared to Windows, the app look to lag "a little bit". It is not slow in any way, it is smooth, but it is more like inertia... For example, in the app (which look like Lunacy) I can drag an element and move it using the mouse, the animation is smooth, but the element lag, may be a quarter of a second, behind the mouse position. It is like if the app recorded all the mouse positions and play them all instead of skiping a few frames to follow the mouse. I did not noticed that behavior on Windows or WASM on Windows.
=> I am totally noob in MacOS so maybe it is related to the trackpad behavior, or may be it is because I run the app from VS. Are you aware of such behavior on Mac system? May be that can be fixed by runing the app outside VS? or compiling in AOT? From what I read in the doc, I need a developer account to sign the app to be able to execute it outside VS right? (I don't have the dev account yet so I did not tried).
=> Or may be there is a settings to configure?
=> **EDIT**: I tried to use a mouse instead of trackpad, and the lag feeling is far less or not noticeable... In fact, this is strange, with mouse: I add a few images to my "document", move one and it will lag a little, release mouse or stop moving, move it again and now it is fast... so may be it is because of VS... because I have no background task or whatever...Or may be Avalonia or Skia are caching some stuffs? With trackpad, it lag more or less all the time.

2) [Bug?] Not a big deal but I noticed that the "size" cursors I am using are wrong on MacOS, I have to check exactly what are the name of the cursors I am using, but they were fine on Windows, and WASM (at least on Windows).

3) [Bug?] I have had an issue with flyouts (combobox, button flyout, etc...), they were not working, I tried many controls (during may be 20 sec) and nothing... but after some time, they were working back normally. Is there a know issue about that?

4) The app (or a sample app) freeze when I try to close it with top left red button (when running from VS), and I have to cancel Debug from Visual to actually close it. Is it normal?

5) Is AOT compilation working on MacOS with .NET7 and does it improve performances?

6) Could you confirm that I need to sign the app with a developer account to try it outside of VS? or may be I missunderstood the doc?

That's all, which is very encouraging for a first compilation! Thanks a lot!

P.S. I am using 11 RC1.1

**Comment by maxkatz6** [comment, marked as answer]
Hi!

1. It's a known issue with a trackpad. Avalonia doesn't group input events on macOS and that causes this visual imperfection there. Windows and WASM input events are grouped by the platform itself for the record, for example.
2. Is it about standard cursors or bitmap ones? With standard cursors there shouldn't be any difference from the native OS. Not sure about bitmap, there are higher chances to have a bug.
3. It wasn't reported before, and in general it seems to work just fine on macOS. Can you create an issue with more information and minimal steps to repro?
4. Not normal, but caused by the debugger. See https://github.com/AvaloniaUI/Avalonia/issues/11159.
5. In .NET AOT is less about performance, but more about `Startup time / RAM / app size` metrics. AOT in general greatly improves startup time, as it removes JIT compilation. NativeAOT works on macOS, but IIRC AOT compilation for ARM was added only in .NET 8. Better to double-check MS documentation for that.
6. Yes. You need. You can run an unsigned app on the same machine, but to run it on other devices you really should sign it. Ideally, even notarize it with an apple dev account, but it's not yet a requirement if you don't distribute app through the app store. If you do - then you also need to sandbox your app (which is yet another headache in general).

**Comment by dbriard** [comment]
Thanks @maxkatz6 for the fast reply!

1. That make sense. I suppose nothing is planned to avoid this behavior in Avalonia? May be I can filter events myself. Is there an equivalent to this WPF event in Avalonia???: CompositionTarget.Rendering

2. It is about standard cursors: I am using those one on Windows:
```
if (name == "Resize_1.cur")
                        cursors[num4] = new Cursor(StandardCursorType.LeftSide);
                    else if (name == "Resize_2.cur")
                        cursors[num4] = new Cursor(StandardCursorType.TopLeftCorner);
                    else if (name == "Resize_3.cur")
                        cursors[num4] = new Cursor(StandardCursorType.TopSide);
                    else if (name == "Resize_4.cur")
                        cursors[num4] = new Cursor(StandardCursorType.TopRightCorner);
```
And they do not look the same on MacOS.. For example LeftSide and TopSide are symetric arrow on Windows, but not on MacOS. I should use all 4 sides instead. TopLeftCorner is diagonal symetric arrow on Windows, while it is a very tiny cross on MacOS.. I suppose the best for cross platform is to use only custom cursors.
![IMG_20230628_084643](https://github.com/AvaloniaUI/Avalonia/assets/3595443/924131eb-749a-42b1-b965-b22959bf484a)

![IMG_20230628_084701](https://github.com/AvaloniaUI/Avalonia/assets/3595443/b0cad52a-b8f7-4ca7-b17c-c517e2a2a2d3)
![IMG_20230628_085000](https://github.com/AvaloniaUI/Avalonia/assets/3595443/aa42e47d-f5dc-4f0f-8b88-07488561adb2)
![IMG_20230628_084954](https://github.com/AvaloniaUI/Avalonia/assets/3595443/883b3f81-796f-4b2a-a803-c436d0b9e63e)

3. I try again and did not reproduced yet, but I will let you know if I reproduce. May be I did something that broke all flyouts, and something else that resolve it. 

4. Thanks, I'll see if latest nightly build resolve that.

5. ok, so not a priority for me at the moment.

6. Good to know, I just tried using Dotnet Bundle steps in the doc, but it do not work. The build work fine and the app is created in macos-x64/publish folder. Unfortunatly when I double click, I got an error "the app is damaged or incomplete." I skipped the chmod step as I am building directly on the Mac. Do you know what may cause this error, or if a step is missing in the doc? Thank you.
**EDIT: I finally managed to bundle the app. I found a tutorial that use a tool called Platypus which have a GUI** :) I will have to check the differences with dotnet msbuild bundle...
